### ISSUE TITLE
[future] Allow loading different versions of the same assembly

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Starcounter#3872]
Currently starcounter doesn't allow loading two assemblies with the same display name, even if their strong names are different. This is because of two reasons:

1. Database objects are referred to by dispaly name of assembly (e.g. `Simplified.Ring1.Something`) - this will hopefully be alleviated once new mapping is in place. This also pertains only to assemblies that introduce database classes (which excludes most of 3rd party code)
1. Starcounter simply refuses to load an assembly with a display name that already has been loaded. This can probably be fixed, since .NET runtime allows it

Can this be really enabled? What are possible obstacles? What are possible benefits?
CC @miyconst @per-samuelsson

**Comment by chrhol** [comment]
For the second point, this is an old problem that surfaces from time to time, and we have yet to solve it correctly. I actually bumped an old issue a couple of days ago since this might be a problem when I push the change in which jsonparser we use. 

It would be really good to raise the priority on this and finally solve it.

Old issues: #1601, #1620

**Comment by miyconst** [comment]
+1.

From HO chat by @joozek78:

Some more readings and a reason to support `app.config` file in Starcounter - [Referencing 2 different versions of log4net in the same solution](http://stackoverflow.com/questions/3158928/referencing-2-different-versions-of-log4net-in-the-same-solution/3163050#3163050).

**Comment by k-rus** [comment]
Isn't it a source for potential problems when it is unclear which actual assembly is used? Such flexibility is usually difficult to debug.

**Comment by joozek78** [comment]
I agree it can be confusing, but sometimes it's necessary. Imagine two
completely independent 3rd party apps that use the same library (let's say
log4net or FluentValidation), but in different versions. Currently we can't
even run them together.

**Comment by per-samuelsson** [comment]
FWIW, we are going towards a model where we can support multiple hosts / database. With such ability, we can ask to load apps that necessarily need this to run in independent hosts. Maybe we should push this until we have that?

Anyone interested in looking at the loading strategies we employ today in more detail, the code host [AssemblyResolver](https://github.com/Starcounter/Starcounter/blob/develop/src/Hosting/Starcounter.Hosting/AssemblyResolver.cs) is the place to look.

**Comment by per-samuelsson** [comment]
> Anyone interested in looking at the loading strategies we employ today in more detail, the code host [AssemblyResolver](https://github.com/Starcounter/Starcounter/blob/develop/src/Hosting/Starcounter.Hosting/AssemblyResolver.cs) is the place to look.

And [this method](https://github.com/Starcounter/Starcounter/blob/develop/src/Hosting/Starcounter.Hosting/AssemblyResolver.cs#L154) would probably be of most interest.
